group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java

```diff
@@ -761,6 +776,58 @@ public ClassicGroup classicGroup(
         }
     }
 
+    /**
+     * Validates the online upgrade if the Classic Group receives a ConsumerGroupHeartbeat request.
+     *
+     * @param classicGroup A ClassicGroup.
+     * @return the boolean indicating whether it's valid to online upgrade the classic group.
+     */
+    private boolean validateOnlineUpgrade(ClassicGroup classicGroup) {
+        if (!consumerGroupMigrationPolicy.isUpgradeEnabled()) {
+            log.debug("Online upgrade is invalid because the consumer group {} migration config is {} so online upgrade is not enabled.",
+                classicGroup.groupId(), consumerGroupMigrationPolicy);
+            return false;
+        } else if (classicGroup.isInState(DEAD)) {
+            log.debug("Online upgrade is invalid because the classic group {} is in DEAD state.", classicGroup.groupId());
+            return false;
+        } else if (!classicGroup.usesConsumerGroupProtocol()) {
+            log.debug("Online upgrade is invalid because the classic group {} has protocol type {} and doesn't use the consumer group protocol.",
+                classicGroup.groupId(), classicGroup.protocolType().orElse(""));
+            return false;
+        } else if (classicGroup.size() > consumerGroupMaxSize) {
+            log.debug("Online upgrade is invalid because the classic group {} size {} exceeds the consumer group maximum size {}.",
+                classicGroup.groupId(), classicGroup.size(), consumerGroupMaxSize);
+            return false;
+        }
+        return true;
+    }
+
+    /**
+     * Creates a ConsumerGroup corresponding to the given classic group.
+     *
+     * @param classicGroup  The ClassicGroup to convert.
+     * @param records       The list of Records.
+     * @return  The created ConsumerGroup.
+     */
+    ConsumerGroup convertToConsumerGroup(ClassicGroup classicGroup, List<Record> records) {
+        // The upgrade is always triggered by a new member joining the classic group, which always results in
+        // updatedMember.subscribedTopicNames changing, the group epoch being bumped, and triggering a new rebalance.
+        // If the ClassicGroup is rebalancing, inform the awaiting consumers of another ongoing rebalance
+        // so that they will rejoin for the new rebalance.
+        classicGroup.completeAllJoinFutures(Errors.REBALANCE_IN_PROGRESS);
+        classicGroup.completeAllSyncFutures(Errors.REBALANCE_IN_PROGRESS);
+
+        classicGroup.createGroupTombstoneRecords(records);
+        ConsumerGroup consumerGroup = new ConsumerGroup(snapshotRegistry, logContext, classicGroup.groupId(), metrics);
+        consumerGroup.fromClassicGroup(classicGroup, records, metadataImage.topics());
```

Comment:
nit: I may be better to have a static method - `ConsumerGroup.fromClassicGroup(....)` - which creates the `ConsumerGroup` and populates it from the classic group.